AllSimplexGenerator mesh generator

roystgnr opened the pull request:
Closes #28833

## Reason
Tet meshes are more flexible than hex/prism meshes, especially when we want to be able to stitch them to tet mesh generator output, but we don't yet have a MOOSE interface to the libMesh all_tri() function that splits non-simplex elements into simplices.

## Design
Add a new AllSimplexGenerator mesh generator; this will just call all_tri() on its input.

## Impact
No change to existing APIs, just a new mesh generator option.

roystgnr committed:
Add AllSimplexGenerator mesh generator

The 2D all_tri() code has been in libMesh forever, but it's been
improved (support for Hexes) recently, and it's going to be more useful
in a MOOSE context (interfacing to tet generation) in the near future,
so let's add an interface for it.

Refs #28833

roystgnr committed:
Add AllSimplexGenerator documentation

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/2491802/), step Docs: sync website on 05648d3 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/28834/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/2491850/), step Generate coverage on 05648d3 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>45387e</th>
      <th colspan=4>#28834 05648d</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.05%</td>
      <td>85.04%</td>
      <td>-0.01%</td>
      <td>30.77%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>106062</td>
      <td>106066</td>
      <td>+4</td>
      <td>4</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>18646</td>
      <td>18655</td>
      <td>+9</td>
      <td>9</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/28834/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/28834/coverage/framework)


## Modules coverage

Coverage did not change

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/28834/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/28834/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/28834/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/28834/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/28834/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/28834/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/28834/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/28834/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/28834/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/28834/coverage/geochemistry)
- [`heat_transfer`](https://mooseframework.inl.gov/docs/PRs/28834/coverage/heat_transfer)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/28834/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/28834/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/28834/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/28834/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/28834/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/28834/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/28834/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/28834/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/28834/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/28834/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/28834/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/28834/coverage/scalar_transport)
- [`solid_mechanics`](https://mooseframework.inl.gov/docs/PRs/28834/coverage/solid_mechanics)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/28834/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/28834/coverage/stochastic_tools)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/28834/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/28834/coverage/xfem)

</details>

## Warnings

- `framework` new line coverage rate 30.77% is less than the suggested 90.0%



This comment will be updated on new commits.

GiudGiud commented on framework/doc/content/source/meshgenerators/AllSimplexGenerator.md:
```diff
@@ -0,0 +1,24 @@
+In 3D, the geometry of any non-planar quadrilateral faces of cubes
+and/or prisms will not be exactly preserved, due to the change from
+bilinear or biquadratic mapping on quadrilaterals to linear or
+quadratic mapping on the triangles that replace them.
```
```suggestion
quadratic mapping on the triangles that replace them.

!syntax parameters /Mesh/AllSimplexGenerator

!syntax inputs /Mesh/AllSimplexGenerator

!syntax children /Mesh/AllSimplexGenerator
```

roystgnr committed:
Update framework/doc/content/source/meshgenerators/AllSimplexGenerator.md

Co-authored-by: Guillaume Giudicelli <guillaume.giudicelli@gmail.com>

GiudGiud commented on framework/doc/content/source/meshgenerators/AllSimplexGenerator.md:
```diff
@@ -0,0 +1,30 @@
+# AllSimplexGenerator
+
+!syntax description /Mesh/AllSimplexGenerator
```
I have a slight preference for Converter over Generator considering the [ElementsToTetrahedronsConverter](https://mooseframework.inl.gov/source/meshgenerators/ElementsToTetrahedronsConverter.html)

prefer AllSimplex... or ElementsToSimplicesConverter ? or SimplexConversionGenerator?
some ideas. We have not been naming MGs after the libMesh all_do_something routines 

which btw we could deprecate?

roystgnr committed:
Fix for non-Unity builds

roystgnr committed:
AllSimplexGenerator->ElementsToSimplicesConverter

This is clearer and fits in better.

roystgnr committed:
Add class description

roystgnr committed:
Regression tests for ElementsToSimplicesConverter

A few of these just symlink the existing gold files from
ElementsToTetrahedronsConverter tests.

We can't reuse the HEX8 gold from ElementsToTetrahedronsConverter
because the splitting rules in that code differ from libMesh's.  libMesh
also doesn't yet have pyramid support (so we definitely can't deprecate
the other converter).  We can add more tests for a couple higher-order
elements that libMesh supports.

roystgnr committed:
Add missing symlink
